draft-ietf-quic-v2.md

```diff
@@ -117,19 +117,40 @@ nonce = 0x141b99c239b03e785d6a2e9f
 ~~~
 
 
-# Version Negotiation Considerations
+# Compatible Version Negotiation
 
 QUIC version 2 endpoints SHOULD also support QUIC version 1. Any QUIC endpoint
 that supports multiple versions MUST fully implement {{QUIC-VN}} to prevent
 version downgrade attacks.
 
-Note that version 2 meets that document's definition of a compatible version
-with version 1. Therefore, v2-capable servers MUST use compatible version
-negotiation unless they do not support version 1.
-
-As version 1 support is more likely than version 2 support, a client SHOULD use
-QUIC version 1 for its original version unless it has out-of-band knowledge that
-the server supports version 2.
+QUIC version 2 is symmetrically compatible with QUIC version 1 as defined in
+{{Section 3 of QUIC-VN}}.  That is, a server can choose to negotiate version 2
+after a client initiates a connection using version 1; or, a server can
+negotiate version 1 if a client initiates a version 2 connection attempt.
+
+A server that supports both version 1 and version 2 SHOULD support a compatible
+upgrade to version 2 if a client offers the option.
+
+As the only difference between version 1 and version 2 is the protection of
+Initial and Retry packets, compatibility is defined as taking the unprotected
+content of those packets and processing them as packets from the negotiated
+version.  Or, considering that endpoints will process packets in either version
+as they are received prior to negotiating a version, the state that is
+accumulated during that processing can be used as though it was the result of
+processing packets from the negotiated version.
+
+A server SHOULD produce a Retry packet in the same version as the Initial packet
+that the Retry packet responds to.  Validation of the Retry exchange uses the
+logic of the final negotiated version, treating the Retry as if it were sent in
+that version.  This means that the version used for Retry is not authenticated,
+only that it occurred and the connection ID that was used.
+
+A server MUST NOT accept 0-RTT when performing a compatible version upgrade.
```

Comment:
Is there a reason for this? It seems like a disincentive to upgrade.

Reply:
Maybe.  Though the number of cases where you get to deploy a whole new version and continue to use 0-RTT across the upgrade is small.  You will mostly have a previous connection in the negotiated version, so you can remember which was used and attempt that straight up.  It is only during the rollout of v2 that you need to worry about maybe accepting v2 and rejecting 0-RTT.

In our stack, it is completely non-trivial to calculate v1 0-RTT keys once you have negotiated v2.  Given how narrow the applicability is, it seems much safer to reject.

Of course, you could just make rejection discretionary.

Reply:
Let see. To Martin's point, I have only partially tested usage of 0-RTT with resumption, always resuming with V1.

A server negotiates 0-RTT if it has received a session resumption ticket during the previous handshake. Session tickets are carried in crypto frames in 1-RTT packets, thus after the transport parameters have been negotiated. It seems we can have the following:

1) Client initiates a V1 session, also proposes V2, server negotiates V2, indicates V2 support, and sends a resumption ticket.

2) Client initiates a V1 session, also proposes V2, server negotiates V2, does not indicates V2 support, and sends a resumption ticket.

3) Client initiates a V1 session, also proposes V2, server negotiates V1 and sends a resumption ticket.

4) Client initiates a V1 session, does not propose alternative. Server sends a resumption ticket.

It seems to me that in case #1, the client should resume a session using V2, including for 0-RTT. In case #2, #3 and #4, it should resume using V1. Case #4 is plain V1, which implies support of 0-RTT. It would be bizarre to allow 0RTT in case #4 but not in cases #2 and #3.

Reply:
I'm not following case 2.  Did you mean that the server negotiates v2, but does not support v1?  You can't use compatible negotiation for that, I think.

Cases 3 and 4 are clear cut cases where the only mutually supported version is v1, just that one has VN support and the other doesn't.

The case that is potentially interesting is this: Client offers v1 and v2.  Server only supports v1 so it uses that and sends a 0-RTT-capable ticket.  Server upgrades to add v2 support.  Client then makes a new connection with v1 and 0-RTT.  Server is forced to choose between a compatible upgrade and 0-RTT.

Reply:
The server response contains a the selected version and a list of supported versions. Case #2 happens when the selected version is V2, but the list of supported versions only contains V1. That would be a reasonable thing to do by a server in a big farm. The server itself supports V2, but not all servers in the farm do, so the server is telling the client that it would be reasonable to use V1 in the new connections.

Reply:
Good point.  I missed that nuance.  That increases the potential value of doing 0-RTT in the original version.